### DISCUSSION TITLE
Is there a way to have app_store_connect submit builds ONLY to internal beta groups?

### DISCUSSION THREAD

**Discussion by VIadislavJevremovic** [codemagic-ci-cd/codemagic-docs#2536] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

iOS (Swift, Objective-C)

### Steps to reproduce

Using this code

```
auth: integration
submit_to_testflight: true
beta_groups:
  - InternalGroup
```

InternalGroup is defined within the 'Internal Testing' section on TestFlight (NOT External).

### Expected results

Builds are sent to InternalGroup WITHOUT triggering Beta App Review.

### Actual results

Builds are sent to InternalGroup triggering Beta App Review.

Quoting App Review: _"We see that this binary has been uploaded for external Testing. If you need your app to be tested by internal testers only, please submit the newer binary for review."_

It is unclear from the documentation page here https://docs.codemagic.io/yaml-publishing/app-store-connect/ whether only external beta group support is supported (appears to be so).

Note that beta groups can be defined as both internal/external and it seems API supports attaching builds to both:
https://developer.apple.com/documentation/appstoreconnectapi/prerelease_versions_and_beta_testers/beta_groups
https://developer.apple.com/documentation/appstoreconnectapi/add_builds_to_a_beta_group

Xcode 15 introduced an option to deliver builds as Internal Only, do you support this?
Having `submit_to_testflight: false` and `beta_groups` defined produces a validation error in the webhook.

Also, why cannot all admin members of a team see the Intercom chat widget on a paid plan? (Safari, no popup blockers).

**Comment by VIadislavJevremovic** [comment]
You specifically require submission to Beta Review when using beta_groups:

```if submit_to_testflight and beta_group_names:```
(https://github.com/codemagic-ci-cd/cli-tools/blob/master/src/codemagic/tools/app_store_connect/actions/publish_action.py#L267)

EDIT: Attaching builds to internal beta groups is indeed possible with App Store Connect API:
```
POST https://api.appstoreconnect.apple.com/v1/builds/BUILD_UUID/relationships/betaGroups
{
    "data": [
        {
            "type": "betaGroups",
            "id": "INTERNAL_BETA_GROUP_ID"
        }
    ]
}
```

@himesh-cm @dtrdic It would be nice if your development team can implement this.

**Reply by dtrdic** [reply]
hey @VIadislavJevremovic, sorry didn't quite understand you, are you saying that you are unable to submit your build to internal testing group on App Store Connect, but only to external testing group? 
Because it's already possible to distribute either on internal or external groups, just make sure you are specifying the right name for your group(it's case sensitive so doublecheck the casings). Do you get any specific error message? A build id would also help
```
        app_store_connect:
          auth: integration 
          submit_to_testflight: true
          beta_groups:
          - internalGroupTestName
```

**Reply by VIadislavJevremovic** [reply]
Hello @dtrdic, yes - we need to upload builds to internal groups in certain workflows but WITHOUT submitting to Testflight beta review. However, if I have `submit_to_testflight: false` and `beta_groups` defined in YAML, the webook complains: `TestFlight submission is required for beta_groups`

In your source code, this seems to be intentional: https://github.com/codemagic-ci-cd/cli-tools/blob/master/src/codemagic/tools/app_store_connect/actions/publish_action.py#L267
Why?

Group names are correct. As a workaround, we're currently submitting the builds with `submit_to_testflight: false` and no `beta_groups` defined. Then we manually assign them, but this is sub-optimal. As I mentioned above, API does allow our intended behavior.

**Reply by dtrdic** [reply]
hey @VIadislavJevremovic the `submit_to_testflight` flag actually controls whether submission goes to TestFlight external beta review or not. So, if `submit_to_testflight` is set to `true`, then you are submitting the app to external groups which requires Apple’s review before.
By default the builds will submitted to the internal group, so simpley remove it from your yaml,  as Codemagic assumes you are submitting to external groups when this flag is present

**Reply by VIadislavJevremovic** [reply]
Hello @dtrdic, thanks for your reply.

What about this message that appears in webhook list: `TestFlight submission is required for beta_groups` if `submit_to_testflight` is set to `false` (or omitted, I guess)? Workflow does not start due to this error.

We need `beta_groups` definition but appears it cannot be set on its own without `submit_to_testflight`(which I've pointed out in your CLI code https://github.com/codemagic-ci-cd/cli-tools/blob/master/src/codemagic/tools/app_store_connect/actions/publish_action.py#L267).

**Reply by dtrdic** [reply]
hey @VIadislavJevremovic have you tried running the build without defining `submit_to_testflight` and `beta_groups`? The default behavior is that builds will be submitted to TestFlight for internal testers. The `submit_to_testflight: true` setting only supposed to work in conjunction with beta_groups and external testers

**Reply by VIadislavJevremovic** [reply]
Hello @dtrdic, yes - as mentioned yesterday (https://github.com/orgs/codemagic-ci-cd/discussions/2536#discussioncomment-9519077) we were doing it as a workaround.
But as we have multiple internal groups, this is sub-optimal, we need to manually assign builds to groups afterwards.

We would only need a scenario where `beta_groups` definition is sufficient to submit to a certain internal group (without mentioning `submit_to_testflight`).
I have forked cli-tools to support this, definitely possible. But support directly in YAML would be better.

**Reply by VIadislavJevremovic** [reply]
Hello again @dtrdic, I'm attaching a screenshot of my patch here. Essentially allowing using beta_groups only without submitting the build for review. This enables us to send test builds to specific internal test groups (manual distribution type). Technically it's merely removing the submit condition and allowing the wait phase.

@priitlatt it would be nice if you could chime in, thanks.

![diff](https://github.com/codemagic-ci-cd/codemagic-docs/assets/156702156/0039e88d-2d54-465c-93d6-51323af88e32)

**Reply by priitlatt** [reply]
@VIadislavJevremovic, looks like you're right in that Apple does not require build to be submitted to TestFlight before it can be added to beta testing groups. I might be mistaken but I recall that when beta groups functionality was added back in '21 (in https://github.com/codemagic-ci-cd/cli-tools/pull/164) Apple required the build to be submitted to TestFlight before group membership could be altered.

**Reply by VIadislavJevremovic** [reply]
Hi @priitlatt, indeed that might have been the case then, I've only recently investigated this API. I did confirm that it works now with endpoint (a few posts above) and with this modified publish action. Now, I might have missed other required changes here in the diff, just hinting at my thought process.

**Comment by helinanever** [comment]
Hi @VIadislavJevremovic, I've submitted your request as an issue to the Codemagic CLI tools repo: https://github.com/codemagic-ci-cd/cli-tools/issues/409. This will be a prerequisite for any changes in Codemagic.

**Reply by VIadislavJevremovic** [reply]
Hi @helinanever thank you, much appreciated.

**Reply by helinanever** [reply, marked as answer]
Hi @VIadislavJevremovic , this change has been now implemented in the latest version of CLI tools. In Codemagic, you can now skip submitting to TestFlight and publish to beta groups directly if you set up a workflow like this one:

```
workflows:
  workflow:
    ...
    publishing:
      app_store_connect:
        auth: integration
        beta_groups:
          - InternalGroup
```
